packages/ckeditor5-typing/src/utils/inlinehighlight.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+import findAttributeRange from './findattributerange';
+
+/**
+ * @module typing/utils/inlinehighlight
+ */
+
+/**
+ * Adds a visual highlight style to an attribute element in which the selection is anchored.
+ * Together with two-step caret movement, they indicate that the user is typing inside the element.
+ *
+ * Highlight is turned on by adding the given class to the attribute element in the view:
+ *
+ * * The class is removed before the conversion has started, as callbacks added with the `'highest'` priority
+ * to {@link module:engine/conversion/downcastdispatcher~DowncastDispatcher} events.
+ * * The class is added in the view post fixer, after other changes in the model tree were converted to the view.
+ *
+ * This way, adding and removing the highlight does not interfere with conversion.
+ *
+ * @param {module:core/editor/editor~Editor} editor The editor instance.
+ * @param {String} attributeName The attribute name to check.
+ * @param {String} tagName The tagName of a view item.
+ * @param {String} className The class name to apply in the view.
+ */
+export default function setupLinkHighlight( editor, attributeName, tagName, className ) {
+	const view = editor.editing.view;
+	const highlightedElements = new Set();
+
+	// Adding the class.
+	view.document.registerPostFixer( writer => {
+		const selection = editor.model.document.selection;
+		let changed = false;
+
+		if ( selection.hasAttribute( attributeName ) ) {
+			const modelRange = findAttributeRange(
+				selection.getFirstPosition(),
+				attributeName,
+				selection.getAttribute( attributeName ),
+				editor.model
+			);
+			const viewRange = editor.editing.mapper.toViewRange( modelRange );
+
+			// There might be multiple view elements in the `viewRange`, for example, when the `a` element is
+			// broken by a UIElement.
+			for ( const item of viewRange.getItems() ) {
+				if ( item.is( tagName ) && !item.hasClass( className ) ) {
```

Comment:
I worry this one :D For `<a>` and `<code>` it should work, but what if someone would downcast code to `<span>`?

It might also, doesn't work for 2SCM enable on attributes which convert to `<span>` - one could do that.

I'll check that.

Reply:
OK, a bit artificial, but possible scenario (might be handled by a follow-up?)

![](https://user-images.githubusercontent.com/247363/87425084-59027900-c5dd-11ea-8766-726fe563d70f.png)

The manual test was updated with this:

```javascript
// Custom attribute downcast (I love spans)
editor.conversion.attributeToElement( {
	model: 'italic', view: {
		name: 'span',
		styles: {
			'color': 'red'
		}
	}, converterPriority: 'high'
} );

editor.conversion.for( 'downcast' ).attributeToElement( {
	model: 'bold',
	view: ( attributeValue, writer ) => {
		if ( !attributeValue ) {
			return;
		}

		return writer.createAttributeElement(
			'span',
			{
				style: 'font-weight:bold'
			},
			{
				priority: 80,
				id: uid()
			} );
	}, converterPriority: 'high'
} );

editor.conversion.for( 'downcast' ).attributeToElement( {
	model: 'underline',
	view: ( attributeValue, writer ) => {
		if ( !attributeValue ) {
			return;
		}

		return writer.createAttributeElement(
			'span',
			{
				style: 'text-decoration:underline'
			},
			{
				priority: 20,
				id: uid()
			} );
	}, converterPriority: 'high'
} );

// Don't mind the function name, but the attribute highlighting must be also re-set.
setupLinkHighlight( editor, 'italic', 'span', 'highlight-i' );
setupLinkHighlight( editor, 'underline', 'span', 'highlight-u' );
```

So the question is - should we care? //cc @Reinmar

Reply:
TL;DR attribute highlighting for 2SCM will not work for spans.

Reply:
Yup, not a problem for now.